// @tanstack/router-devtools-core/dist/esm/Explorer.js
import { template, insert, memo, createComponent, effect, className, setAttribute, mergeProps, delegateEvents } from "solid-js/web";
import { clsx } from "clsx";
import * as goober from "goober";
import { createSignal, createMemo, useContext } from "solid-js";
import { tokens } from "./tokens.js";
import { displayValue } from "./utils.js";
import { ShadowDomTargetContext } from "./context.js";
var _tmpl$ = /* @__PURE__ */ template(`<span><svg xmlns=http://www.w3.org/2000/svg width=12 height=12 fill=none viewBox="0 0 24 24"><path stroke=currentColor stroke-linecap=round stroke-linejoin=round stroke-width=2 d="M9 18l6-6-6-6">`), _tmpl$2 = /* @__PURE__ */ template(`<div>`), _tmpl$3 = /* @__PURE__ */ template(`<button><span> `), _tmpl$4 = /* @__PURE__ */ template(`<div><div><button> [<!> ... <!>]`), _tmpl$5 = /* @__PURE__ */ template(`<button><span></span> 🔄 `), _tmpl$6 = /* @__PURE__ */ template(`<span>:`), _tmpl$7 = /* @__PURE__ */ template(`<span>`);
const Expander = ({
  expanded,
  style = {}
}) => {
  const styles = useStyles();
  return (() => {
    var _el$ = _tmpl$(), _el$2 = _el$.firstChild;
    effect((_p$) => {
      var _v$ = styles().expander, _v$2 = clsx(styles().expanderIcon(expanded));
      _v$ !== _p$.e && className(_el$, _p$.e = _v$);
      _v$2 !== _p$.t && setAttribute(_el$2, "class", _p$.t = _v$2);
      return _p$;
    }, {
      e: void 0,
      t: void 0
    });
    return _el$;
  })();
};
function chunkArray(array, size) {
  if (size < 1) return [];
  let i = 0;
  const result = [];
  while (i < array.length) {
    result.push(array.slice(i, i + size));
    i = i + size;
  }
  return result;
}
function isIterable(x) {
  return Symbol.iterator in x;
}
function Explorer({
  value,
  defaultExpanded,
  pageSize = 100,
  filterSubEntries,
  ...rest
}) {
  const [expanded, setExpanded] = createSignal(Boolean(defaultExpanded));
  const toggleExpanded = () => setExpanded((old) => !old);
  const type = createMemo(() => typeof value());
  const subEntries = createMemo(() => {
    let entries = [];
    const makeProperty = (sub) => {
      const subDefaultExpanded = defaultExpanded === true ? {
        [sub.label]: true
      } : defaultExpanded == null ? void 0 : defaultExpanded[sub.label];
      return {
        ...sub,
        value: () => sub.value,
        defaultExpanded: subDefaultExpanded
      };
    };
    if (Array.isArray(value())) {
      entries = value().map((d, i) => makeProperty({
        label: i.toString(),
        value: d
      }));
    } else if (value() !== null && typeof value() === "object" && isIterable(value()) && typeof value()[Symbol.iterator] === "function") {
      entries = Array.from(value(), (val, i) => makeProperty({
        label: i.toString(),
        value: val
      }));
    } else if (typeof value() === "object" && value() !== null) {
      entries = Object.entries(value()).map(([key, val]) => makeProperty({
        label: key,
        value: val
      }));
    }
    return filterSubEntries ? filterSubEntries(entries) : entries;
  });
  const subEntryPages = createMemo(() => chunkArray(subEntries(), pageSize));
  const [expandedPages, setExpandedPages] = createSignal([]);
  const [valueSnapshot, setValueSnapshot] = createSignal(void 0);
  const styles = useStyles();
  const refreshValueSnapshot = () => {
    setValueSnapshot(value()());
  };
  const handleEntry = (entry) => createComponent(Explorer, mergeProps({
    value,
    filterSubEntries
  }, rest, entry));
  return (() => {
    var _el$3 = _tmpl$2();
    insert(_el$3, (() => {
      var _c$ = memo(() => !!subEntryPages().length);
      return () => _c$() ? [(() => {
        var _el$4 = _tmpl$3(), _el$5 = _el$4.firstChild, _el$6 = _el$5.firstChild;
        _el$4.$$click = () => toggleExpanded();
        insert(_el$4, createComponent(Expander, {
          get expanded() {
            return expanded() ?? false;
          }
        }), _el$5);
        insert(_el$4, () => rest.label, _el$5);
        insert(_el$5, () => String(type).toLowerCase() === "iterable" ? "(Iterable) " : "", _el$6);
        insert(_el$5, () => subEntries().length, _el$6);
        insert(_el$5, () => subEntries().length > 1 ? `items` : `item`, null);
        effect((_p$) => {
          var _v$3 = styles().expandButton, _v$4 = styles().info;
          _v$3 !== _p$.e && className(_el$4, _p$.e = _v$3);
          _v$4 !== _p$.t && className(_el$5, _p$.t = _v$4);
          return _p$;
        }, {
          e: void 0,
          t: void 0
        });
        return _el$4;
      })(), memo(() => memo(() => !!(expanded() ?? false))() ? memo(() => subEntryPages().length === 1)() ? (() => {
        var _el$7 = _tmpl$2();
        insert(_el$7, () => subEntries().map((entry, index) => handleEntry(entry)));
        effect(() => className(_el$7, styles().subEntries));
        return _el$7;
      })() : (() => {
        var _el$8 = _tmpl$2();
        insert(_el$8, () => subEntryPages().map((entries, index) => {
          return (() => {
            var _el$9 = _tmpl$4(), _el$10 = _el$9.firstChild, _el$11 = _el$10.firstChild, _el$12 = _el$11.firstChild, _el$17 = _el$12.nextSibling, _el$14 = _el$17.nextSibling, _el$18 = _el$14.nextSibling;
            _el$18.nextSibling;
            _el$11.$$click = () => setExpandedPages((old) => old.includes(index) ? old.filter((d) => d !== index) : [...old, index]);
            insert(_el$11, createComponent(Expander, {
              get expanded() {
                return expandedPages().includes(index);
              }
            }), _el$12);
            insert(_el$11, index * pageSize, _el$17);
            insert(_el$11, index * pageSize + pageSize - 1, _el$18);
            insert(_el$10, (() => {
              var _c$3 = memo(() => !!expandedPages().includes(index));
              return () => _c$3() ? (() => {
                var _el$19 = _tmpl$2();
                insert(_el$19, () => entries.map((entry) => handleEntry(entry)));
                effect(() => className(_el$19, styles().subEntries));
                return _el$19;
              })() : null;
            })(), null);
            effect((_p$) => {
              var _v$5 = styles().entry, _v$6 = clsx(styles().labelButton, "labelButton");
              _v$5 !== _p$.e && className(_el$10, _p$.e = _v$5);
              _v$6 !== _p$.t && className(_el$11, _p$.t = _v$6);
              return _p$;
            }, {
              e: void 0,
              t: void 0
            });
            return _el$9;
          })();
        }));
        effect(() => className(_el$8, styles().subEntries));
        return _el$8;
      })() : null)] : (() => {
        var _c$2 = memo(() => type() === "function");
        return () => _c$2() ? createComponent(Explorer, {
          get label() {
            return (() => {
              var _el$20 = _tmpl$5(), _el$21 = _el$20.firstChild;
              _el$20.$$click = refreshValueSnapshot;
              insert(_el$21, () => rest.label);
              effect(() => className(_el$20, styles().refreshValueBtn));
              return _el$20;
            })();
          },
          value: valueSnapshot,
          defaultExpanded: {}
        }) : [(() => {
          var _el$22 = _tmpl$6(), _el$23 = _el$22.firstChild;
          insert(_el$22, () => rest.label, _el$23);
          return _el$22;
        })(), " ", (() => {
          var _el$24 = _tmpl$7();
          insert(_el$24, () => displayValue(value()));
          effect(() => className(_el$24, styles().value));
          return _el$24;
        })()];
      })();
    })());
    effect(() => className(_el$3, styles().entry));
    return _el$3;
  })();
}
const stylesFactory = (shadowDOMTarget) => {
  const {
    colors,
    font,
    size
  } = tokens;
  const {
    fontFamily,
    lineHeight,
    size: fontSize
  } = font;
  const css = shadowDOMTarget ? goober.css.bind({
    target: shadowDOMTarget
  }) : goober.css;
  return {
    entry: css`
      font-family: ${fontFamily.mono};
      font-size: ${fontSize.xs};
      line-height: ${lineHeight.sm};
      outline: none;
      word-break: break-word;
    `,
    labelButton: css`
      cursor: pointer;
      color: inherit;
      font: inherit;
      outline: inherit;
      background: transparent;
      border: none;
      padding: 0;
    `,
    expander: css`
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: ${size[3]};
      height: ${size[3]};
      padding-left: 3px;
      box-sizing: content-box;
    `,
    expanderIcon: (expanded) => {
      if (expanded) {
        return css`
          transform: rotate(90deg);
          transition: transform 0.1s ease;
        `;
      }
      return css`
        transform: rotate(0deg);
        transition: transform 0.1s ease;
      `;
    },
    expandButton: css`
      display: flex;
      gap: ${size[1]};
      align-items: center;
      cursor: pointer;
      color: inherit;
      font: inherit;
      outline: inherit;
      background: transparent;
      border: none;
      padding: 0;
    `,
    value: css`
      color: ${colors.purple[400]};
    `,
    subEntries: css`
      margin-left: ${size[2]};
      padding-left: ${size[2]};
      border-left: 2px solid ${colors.darkGray[400]};
    `,
    info: css`
      color: ${colors.gray[500]};
      font-size: ${fontSize["2xs"]};
      padding-left: ${size[1]};
    `,
    refreshValueBtn: css`
      appearance: none;
      border: 0;
      cursor: pointer;
      background: transparent;
      color: inherit;
      padding: 0;
      font-family: ${fontFamily.mono};
      font-size: ${fontSize.xs};
    `
  };
};
function useStyles() {
  const shadowDomTarget = useContext(ShadowDomTargetContext);
  const [_styles] = createSignal(stylesFactory(shadowDomTarget));
  return _styles;
}
delegateEvents(["click"]);
export {
  Expander,
  Explorer,
  chunkArray
};
